<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [4]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [72]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [12]:
# текст запроса
query_3_1 = f'''select 
                count(name)
                from VACANCIES
'''
df = pd.read_sql_query(query_3_1, connection)
df


C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\2321247111.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [13]:
# результат запроса
49197

49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [26]:
# текст запроса
query_3_2 = f'''select 
                count(name)
                from employers
'''
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1639092598.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [27]:
# текст запроса
query_3_3 = f'''select 
                count(name)
                from areas
'''
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\4118013331.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

In [17]:
# результат запроса
1362

1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [28]:
# текст запроса
query_3_4 = f'''select 
                count(name)
                from industries
'''
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1203554374.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

In [19]:
# результат запроса
294

294

Исходя из данных видно, что количество работодателей в двое меньше чем вакансий, из этого следует, что у многих работодателей по 2 и более вакансий.
В каждом регионе есть хотя бы по одному или 2 работодателя

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [45]:
# текст запроса
query_4_1 = f'''select 
                    a.name,
                    count(v.area_id) cnt
                from areas a
                join VACANCIES v on v.area_id=a.id
                group by 1
                order by 2 desc
                limit 5               
'''
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1623968759.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [55]:
# текст запроса
# текст запроса
query_4_2 = f'''select 
                    count(name),
                from VACANCIES
                where salary_from is NOT NULL or salary_to is NOT NULL
'''              
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\4179948098.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

In [56]:
# результат запроса
24073

24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [78]:
# текст запроса
query_4_3 = f'''select 
                    avg (salary_from) as salary_from, 
                    avg (salary_to) as salary_to
                from VACANCIES
                
'''              
df = pd.read_sql_query(query_4_3, connection)
df


C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1341226337.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


salary_from      salary_to
0  71064.657901  110536.741923

In [79]:
# результат запроса
round(df['salary_from']),round(df['salary_to'])

(0    71065.0
 Name: salary_from, dtype: float64,
 0    110537.0
 Name: salary_to, dtype: float64)

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [92]:
# текст запроса
query_4_4 = f'''select 
                    schedule,
                    employment
                from VACANCIES 
                order by 1 desc
                limit 2
                
'''              
df = pd.read_sql_query(query_4_4, connection)
df               

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\89369127.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment
0  Удаленная работа  Полная занятость
1  Удаленная работа  Полная занятость

Удалённая работа — Полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [102]:
# текст запроса
query_4_4 = f'''select 
                    experience,
                    count(name)
                from VACANCIES 
                group by 1
                order by 2 asc
                
                
'''              
df = pd.read_sql_query(query_4_4, connection)
df  

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1336726206.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет

***

В Москве самое наибольшее количество вакансий.
Часто требуемый рабочий график удаленная работа- полная занятость.
В основном требуемый опыт работы более 6 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [109]:
# текст запроса
query_5_1 = f'''select 
                    e.name,
                    count(v.name)
                from VACANCIES v
                join EMPLOYERS e on v.employer_id= e.id
                group by e.id
                order by 2 desc
                limit 5
                
                
'''              
df = pd.read_sql_query(query_5_1, connection)
df  

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\2704263839.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

Яндекс — Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [125]:
query_5_2 = f'''select 
                    a.name,
                    count(e.id) cnt_e,
                    count(v.id) cnt_v
                from areas a
                left join EMPLOYERS e on a.id= e.area
                left join vacancies v on v.area_id=a.id
                where v.id is null
                group by a.id
                
                order by 2 desc
                
                
                
'''              
df = pd.read_sql_query(query_5_2, connection)
df  

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1955895672.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  cnt_e  cnt_v
0                Россия    410      0
1             Казахстан    207      0
2    Московская область     75      0
3    Краснодарский край     19      0
4    Ростовская область     18      0
..                  ...    ...    ...
588             Скидель      0      0
589          Белоозерск      0      0
590           Орловский      0      0
591          Павловская      0      0
592             Таксимо      0      0

[593 rows x 3 columns]

Россия

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [147]:
query_5_3 = f'''select 
                    distinct(e.name),
                    count( distinct v.area_id) CNT_a
                
                from EMPLOYERS e
                join vacancies v on v.employer_id=e.id
                group by e.name
                order by 2 desc
                
                
                
'''              
df = pd.read_sql_query(query_5_3, connection)
df                 

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\2905813211.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  cnt_a
0                       Яндекс    181
1                   Ростелеком    152
2                   Спецремонт    116
3       Поляков Денис Иванович     88
4                    ООО ЕФИН      71
...                        ...    ...
14761     Яроцкий Е. В./ MoonX      1
14762  Ясли-сад №374 г. Минска      1
14763                      ЯСП      1
14764        Ястро-Переработка      1
14765                     ЯТСС      1

[14766 rows x 2 columns]

181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [335]:
query_5_4 = f'''select 
                 
                COUNT(*) cnt 
                from EMPLOYERS e
                left join EMPLOYERS_INDUSTRIES ei on e.id=ei.employer_id
               where ei.industry_id is null
                
                
'''              
df = pd.read_sql_query(query_5_4, connection)
df                                                                                  

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1879312548.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt
0  8419

8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [245]:
query_5_5 = f'''select 
                  e.name,
                  count(i.name) cnt_i
          
                  
                from EMPLOYERS_INDUSTRIES ei
                join EMPLOYERS e on e.id=ei.employer_id
                join INDUSTRIES i on i.id=ei.industry_id
                
                group by  e.name
                HAVING count(i.name)='4'
                order by 1, 2 asc
'''              
df = pd.read_sql_query(query_5_5, connection)
df    

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\3201724009.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  cnt_i
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]

2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [262]:
query_5_6 = f'''select 
                  i.name,
                  count(e.name) cnt_e
          
                  
                from EMPLOYERS_INDUSTRIES ei
                join EMPLOYERS e on e.id=ei.employer_id
                join INDUSTRIES i on i.id=ei.industry_id
                where i.name = 'Разработка программного обеспечения'
                group by  i.name
                
                
'''              
df = pd.read_sql_query(query_5_6, connection)
df  

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\2097023207.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


name  cnt_e
0  Разработка программного обеспечения   3553

3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников

In [277]:
query_5_7 = f'''select 
                    a.name,
                    count(v.id) cnt_V
                FROM vacancies v
                LEFT JOIN employers e on v.employer_id=e.id
                LEFT JOIN areas a on v.area_id=a.id
                where e.name= 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург','Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа' ,'Ростов на-Дому', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')    
                group by a.name
                order by 2 desc
                
                UNION ALL
                select 'total',sum(cnt_V)    
                from vacancies
                order by 2 desc
                
'''              
df = pd.read_sql_query(query_5_7, connection)
df                 

C:\Users\yanas\AppData\Local\Temp\ipykernel_2800\1018175970.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


DatabaseError: Execution failed on sql 'select 
                    a.name,
                    count(v.id) cnt_V
                FROM vacancies v
                LEFT JOIN employers e on v.employer_id=e.id
                LEFT JOIN areas a on v.area_id=a.id
                where e.name= 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург','Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа' ,'Ростов на-Дому', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')    
                group by a.name
                order by 2 desc
                
                UNION ALL
                select 'total',sum(cnt_V)    
                from vacancies
                order by 2 desc
                
': syntax error at or near "UNION"
LINE 11:                 UNION ALL
                         ^


в модуле не нашла информацию, как правильно вставлять  UNION ALL в пайтон. в метабазе все получается

***

у Большинства работодателей не указана сфера деятельности, что у сложнит поиск для соискателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [87]:
query_6_1 = f'''select 
                count( v.name)
                
                
                from VACANCIES v
                where (lower (v.name) LIKE '%data%')
                    or (lower(v.name) LIKE '%данн%')
                   
                           
'''              
df = pd.read_sql_query(query_6_1, connection)
df    

C:\Users\yanas\AppData\Local\Temp\ipykernel_11488\1078697537.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

Целый день убить из-за этой задачи. И все из-за не коректности задания(нужны ли здесь только уникальные названия или все и т.п.)

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [88]:
query_6_2 = f'''select
                    
                    count(v.id)
                from vacancies v
                where ((lower(v.name) LIKE '%data scientist%')
                    or (lower(v.name) LIKE '%ata science%')
                    or (lower(v.name) LIKE '%исследователь данных%')
                    or ((v.name LIKE '%ML%') AND (lower(v.name) NOT LIKE '%html%'))
                    or (lower(v.name) LIKE '%machine learning%')
                    or (lower(v.name) LIKE '%машинн%обучен%'))
                    and ((lower(v.name) LIKE '%junior%') 
                    or (lower(v.experience) LIKE '%нет опыта%') 
                    or (lower(v.employment) LIKE '%стажировка%'))
                    
                
'''              
df = pd.read_sql_query(query_6_2, connection)
df                

C:\Users\yanas\AppData\Local\Temp\ipykernel_11488\2678652921.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

In [89]:
51

51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [92]:
query_6_3 = f'''select
                    count(v.id)
                from vacancies v
                where ((lower(v.name) LIKE '%data scientist%')
                    or (lower(v.name) LIKE '%ata science%')
                    or (lower(v.name) LIKE '%исследователь данных%')
                    or ((v.name LIKE '%ML%') AND (lower(v.name) NOT LIKE '%html%'))
                    or (lower(v.name) LIKE '%machine learning%')
                    or (lower(v.name) LIKE '%машинн%обучен%'))
                    and ((lower(v.key_skills) LIKE '%sql%') 
                    or (lower(v.key_skills) LIKE '%postgres%'))
                   
                        
               
'''              
df = pd.read_sql_query(query_6_3, connection)
df                          

C:\Users\yanas\AppData\Local\Temp\ipykernel_11488\2601620914.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

201 вакансия 

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [95]:
query_6_4 = f'''select
                    
                    count(v.id)
                from vacancies v
                where ((lower(v.name) LIKE '%data scientist%')
                    or (lower(v.name) LIKE '%ata science%')
                    or (lower(v.name) LIKE '%исследователь данных%')
                    or ((v.name LIKE '%ML%') AND (lower(v.name) NOT LIKE '%html%'))
                    or (lower(v.name) LIKE '%machine learning%')
                    or (lower(v.name) LIKE '%машинн%обучен%'))
                    and (lower(v.key_skills) LIKE '%python%') 
'''              
df = pd.read_sql_query(query_6_4, connection)
df 

C:\Users\yanas\AppData\Local\Temp\ipykernel_11488\142122896.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

351 вакансия

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [99]:
query_6_5 = f'''select
                    ROUND(avg(length (v.key_skills)-
                    length(replace(v.key_skills, CHR(9),''))+1),2)
                from vacancies v
                where ((lower(v.name) LIKE '%data scientist%')
                    or (lower(v.name) LIKE '%ata science%')
                    or (lower(v.name) LIKE '%исследователь данных%')
                    or ((v.name LIKE '%ML%') AND (lower(v.name) NOT LIKE '%html%'))
                    or (lower(v.name) LIKE '%machine learning%')
                    or (lower(v.name) LIKE '%машинн%обучен%'))
                    and v.key_skills is not null
'''                    
df = pd.read_sql_query(query_6_5, connection)
df 

C:\Users\yanas\AppData\Local\Temp\ipykernel_11488\3103294413.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6.406032= 6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [102]:
query_6_6 = f'''select
                    ROUND(avg( coalesce((v.salary_from + v.salary_to)/2,
                    v.salary_from , v.salary_to)),0)
                from vacancies v
                where ((lower(v.name) LIKE '%data scientist%')
                    or (lower(v.name) LIKE '%ata science%')
                    or (lower(v.name) LIKE '%исследователь данных%')
                    or ((v.name LIKE '%ML%') AND (lower(v.name) NOT LIKE '%html%'))
                    or (lower(v.name) LIKE '%machine learning%')
                    or (lower(v.name) LIKE '%машинн%обучен%'))
                    and (experience= 'От 3 до 6 лет')
                   group by v.experience
'''                    
df = pd.read_sql_query(query_6_6, connection)
df 

C:\Users\yanas\AppData\Local\Temp\ipykernel_11488\301852864.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


round
0  243115.0

Заработная плата для ДС составит 243115 руб с учетом опыта работы от 3_6 лет

***

существуют вакансии для начинающий ДС-ов 51 вакансия из общего числа для ДС 1771.
Для опытных специалистов ДС довольно высокий уровень ЗП

# Общий вывод по проекту

Если смотреть на общее число вакансий и вакансии для ДС,то в целом можно сказать,что  из 27 вакансий будет хотя бы одна - ДС.

В горордах миллиониках таких вакансии встречаются чаще. Москва самый распространный город с вакансиями ДС, с условием удаленной работы и приличной заработной платой.